syntax = "proto2";
enum RequestType {
  PING = 0;
  SUBSCRIBE = 1;
  DOWNLOAD = 2;
  UPLOAD = 3;
  LIST = 4;
  LOGIN = 5;
  LOGOUT = 6;
  FILE_UPDATE = 7;
}

message Request {
  required RequestType type = 1;
  // Login
  optional string username = 2;
  optional string password = 3;
  // Download / Upload
  optional string filename = 4;
  // Upload
  optional string data = 5;

  // File Update
  optional FileUpdate file_update = 6;

}

enum ResponseType {
  LOGIN_OK = 0;
  ERROR = 1;
  DATA = 2;
  UPDATED = 3;
  FILE_LIST = 4;
  PONG = 5;
  FILE_UPDATED = 6;
}

message Response {
  required ResponseType type = 1;
  // ERROR
  optional string error_message = 4;
  // UPLOAD | DOWNLOAD | LIST | UPDATED
  optional string error_msg = 2;
  // LOGIN
  optional int32 session_id = 3;
  // FILE SUBSCRIPTION UPDATED
  optional FileUpdate file_update = 5;
  // FILE LIST
  optional FileList file_list = 6;
}

message FileDescriptor {
  required string filename = 1;
}

message FileList {
  repeated FileDescriptor files = 1;
}

message FileData {
  required FileDescriptor file = 1;
  required string data = 2;
}

message FileUpdate {
  required string filename = 1;
  required bool deleted = 2;
  required bytes data = 3;
  required bytes hash = 4;
}